### PR TITLE
[WIP] Various updates to GCC package

### PR DESCRIPTION
**Pull request by adamjstewart**
The main goal of this PR is to remove the `binutils` dependency from the GCC package. Every time I've ever built `gcc+binutils` (the default on Linux) it has crashed. Other users have reported the same thing.

From what I understand, binutils was added as a dependency because someone tried building GCC on a system with an ancient version of binutils. This problem is already documented in Spack's [Getting Started](http://spack.readthedocs.io/en/latest/getting_started.html#binutils) section. According to the documentation, `binutils` shouldn't be added as a dependency, mainly because every package containing C/C++/Fortran code _technically_ depends on `binutils`, but we don't want to add it to every package. Instead, users who run into this problem should build a newer version of binutils and add it to their environment manually. It's more of a Spack dependency than a GCC dependency.

Aside from binutils, this PR also updates GCC to use the new `AutotoolsPackage` base class and removes old hacks that are no longer needed. I marked this WIP because it should probably be tested on a few different OSes (mainly macOS) before it is merged.

### CODE REVIEW

File: var/spack/repos/builtin/packages/gcc/package.py [hunk 1]
```diff
@@ -60,93 +74,82 @@ class Gcc(Package):
     patch('piclibs.patch', when='+piclibs')
     patch('gcc-backport.patch', when='@4.7:5.3')
 
-    def install(self, spec, prefix):
-        # libjava/configure needs a minor fix to install into spack paths.
-        filter_file(r"'@.*@'", "'@[[:alnum:]]*@'", 'libjava/configure',
-                    string=True)
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/gcc/package.py]: I removed this thing. I'm pretty sure this was added back in the day when the Spack installation directories still included special characters like `@`, but I don't think this is necessary anymore.

**Comment by tgamblin** [comment]
@mplegendre: I think you will want to look at this one.

**Comment by tgamblin** [comment]
@mplegendre have some ideas about how we can make Spack's dependency management for compilers more robust, and that would include things like `libc`, runtime libraries, and the `ld` implementation.  I am nervous about taking `binutils` out because it provides `ld`, and we'd eventually like that to be swappable (e.g. maybe we want to provide a version that uses `gold`).  I am not sure how much we actually rely on the `gcc` binutils or whether that's ever really worked.  I suspect we're getting by with the system `ld` most of the time, but *that may not be enough on some systems*.

So I think the goal here should be to figure out whether `binutils` should be removed or just disabled for a while, and whether there are actually use cases that need it, before we decide to obliterate it.

@mplegendre: correct me if I'm wrong here.

**Comment by adamjstewart** [comment]
That makes sense, I can see why we would make an exception for a compiler. I'm fine with either:

1. Setting binutils to default to False
2. Fixing the GCC build with binutils

I'm willing to do the former in this PR. Maybe someone else could handle the latter in another PR?

**Comment by citibeth** [comment]
I never had a problem with GCC+binutils.  Worked for me on Ubuntu 12&14, CentOS7, SuSE Enterprise 11.

I too am nervous about changes to this package.  GCC is huge and takes a long time to build, so I never rebuild it.  We will have to be intentional about testing changes to this, lest we introduce bugs that aren't noticed for a long time.

This should be tested on a lot of platforms.  On a positive note... GCC doesn't have many dependencies, so there really aren't many combinatorial versions to test.

**Comment by alalazo** [comment]
TL;DR I think here we are removing the lowest brick in spack Jenga. Without adding features or solving problems. And it should be tested on MANY different OSes before it gets merged. 

Four reasons for the thumb down : 

 1. I remember the amount of time spent last year on making `binutils` work and the generated `gcc` be compiled on as many platforms as possible : there are reasons for all the flags in there, and I would check the relevant commits carefully before removing them.
 1. as @citibeth mentioned, this package never failed me since then.
 1. the solution shouldn't be to use `--dirty` and get done with it. That flag is a hack to get people going or to work around cases where there's no solution or we don't have one yet. Overly advertising its use will be counter-productive in the long run. 
 1. the link in the docs is true for normal `package`s, but this one is a compiler. `binutils` can even be bootstrapped with GCC.

**Comment by alalazo** [comment]
> I never had a problem with GCC+binutils. Worked for me on Ubuntu 12&14, CentOS7, SuSE Enterprise 11.

rhel6 and rhel7....

**Comment by davydden** [comment]
It successfully installed GCC on Ubuntu 16 and macOS Sierra.

File: var/spack/repos/builtin/packages/gcc/package.py [hunk 2]
```diff
@@ -60,93 +74,82 @@ class Gcc(Package):
...
+    def build(self, spec, prefix):
+        if sys.platform == 'darwin':
+            make('bootstrap')
+        else:
+            make()
+
+    def install(self, spec, prefix):
+        make('install')
 
         self.write_rpath_specs()
```
Comment (alalazo) [inline var/spack/repos/builtin/packages/gcc/package.py]: OT: in special packages like this one, if you need to run something after a default phase, you can use decorators : 
```python
# def install(self, spec, prefix): No need for that if it doesn't change

@AutotoolsPackage.sanity_check('install')
def write_rpath_specs(self):
   ...
```

Comment (adamjstewart) [inline var/spack/repos/builtin/packages/gcc/package.py]: I was pretty sure there was a way to do that but wasn't sure about the syntax, thanks!

Comment (alalazo) [inline var/spack/repos/builtin/packages/gcc/package.py]: My fault : those things are well documented (or at least this was my intention) in the code, but they didn't find their way through the "packaging" section of the docs yet

**Comment by adamjstewart** [comment]
It looks like there's significant push-back against this PR, so I'll close for now. But keep in mind that many users are having problems getting `gcc+binutils` to build:

@svenevs reported this in https://github.com/LLNL/spack/issues/2044#issuecomment-255620335
@willfurnass reported this in https://github.com/LLNL/spack/issues/1866#issuecomment-253181700
@eschnett and @chvillanuevap reported problems with `+binutils` on macOS in https://github.com/LLNL/spack/issues/1487#issuecomment-241406475 and #38, but that is fixed now thanks to #364.

The problem may be related to @mplegendre's comment on https://github.com/LLNL/spack/pull/121#issuecomment-150014993 which links to the following compatibility chart:
http://wiki.osdev.org/Cross-Compiler_Successful_Builds
Maybe we should start hardcoding these binutils version dependencies. But that will get complicated very quickly.

**Comment by citibeth** [comment]
Yes, I think that hard-coding GCC/binutils dependencies would be worthwhile.

**Comment by mplegendre** [comment]
Adding a tighter "compatibility matrix" for bintuils/gcc versions would be nice, but I'm skeptical that it would fix the problems people are seeing with gcc+binutils.  In my experience there can be problems using an old binutils with a newer gcc--particularly when the compiler starts using newer options or relocation types that the linker doesn't understand.  But I don't remember ever seeing problems with new binutils and older gcc.  And a typical gcc+binutils should just install the newest binutils, so I'd expect it to work.

It would be great if we could get some in-depth debugging on some of these gcc+binutils issues to guide us.  I'm not seeing in a detailed breakdown of why gcc~binutils was needed in any of the issues @adamjstewart linked to.

One possibility problems when two binutils installs get mixed together (e.g., libraries from one version and executables from another).  Spack's modification of the gcc spec file should prevent this, but perhaps we're missing some cases.

**Comment by adamjstewart** [comment]
For the record, #2301 is another case of `gcc+binutils` breaking, but `gcc~binutils` compiling successfully.
